opencti-platform/opencti-graphql/src/utils/stix-filtering/boolean-logic-engine.ts

```diff
@@ -0,0 +1,187 @@
+import type { Filter, FilterGroup } from './filter-group';
+
+type FilterLogic = Pick<Filter, 'mode' | 'operator'>;
+type FilterExcerpt = Pick<Filter, 'mode' | 'operator' | 'values'>;
+
+/**
+ * The Boolean Logic Engine is responsible for testing some data recursively against a filter group.
+ * The model of the data is unknown (specifically, we are not using any stix concept here).
+ * The engine job is to compare strings, booleans, numbers and dates with a nested AND/OR logic.
+ */
+
+/**
+ * Utility function that takes a single value and return an array.
+ * The array contains the value only if the value is defined, otherwise array is empty
+ */
+export const toValidArray = <T = unknown>(v: T) => {
+  if (v !== undefined && v !== null) {
+    return [v];
+  }
+  return [];
+};
+
+/**
+ * Apply the filtering logic on values that are compatible with simple arithmetic operators (string, number, boolean).
+ *  - With operator gt, gte, lt or lte, string values are compared alphabetically.
+ *  - With boolean values, expect these operators to work as if true is 1 and false 0 (i.e. true > false).
+ * @param filter the filter with mode and operator
+ * @param adaptedFilterValues filter.values (strings) adapted for the test (e.g. parsing numbers, forcing lower case...)
+ * @param stixCandidates the values inside the DATA that we compare to the filter values; they are properly types
+ *                       We always assume an array of value(s) ; use toValidArray if the data is a single, nullable value.
+ */
+export const testGenericFilter = <T extends string | number | boolean>({ mode, operator }: FilterLogic, adaptedFilterValues: T[], stixCandidates: T[]) => {
+  // "(not) nil" or "(not) equal to nothing" is resolved the same way
+  if (operator === 'nil' || (operator === 'eq' && adaptedFilterValues.length === 0)) {
+    return stixCandidates.length === 0;
+  }
+  if (operator === 'not_nil' || (operator === 'not_eq' && adaptedFilterValues.length === 0)) {
+    return stixCandidates.length > 0;
+  }
+
+  // excluding the cases above, comparing to nothing is not supported and would never match
+  if (adaptedFilterValues.length === 0) {
+    return false;
+  }
+  if (mode === 'and') {
+    // we need to find all of them or none of them
+    return (operator === 'eq' && adaptedFilterValues.every((v) => stixCandidates.includes(v)))
+      || (operator === 'not_eq' && adaptedFilterValues.every((v) => !stixCandidates.includes(v)))
+
+      // In real cases, there is only 1 filter value with the next operators (not much sense otherwise)
+      || (operator === 'lt' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c < v)))
+      || (operator === 'lte' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c <= v)))
+      || (operator === 'gt' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c > v)))
+      || (operator === 'gte' && adaptedFilterValues.every((v) => stixCandidates.some((c) => c >= v)));
```

Comment:
What about replacing it by 
```suggestion
      || (['lt', 'lte', 'gt', 'gte'].includes(operator) && adaptedFilterValues.every((v) => stixCandidates.some((c) => c < v)))
```

Reply:
there are all different comparison operators, why only use `c < v` ? it's only for "lower than"

Reply:
I tried to read it many time but did not saw it.